components/net/http_loader.rs

```diff
@@ -1251,7 +1252,81 @@ fn http_network_or_cache_fetch(
     // TODO: if necessary set response's range-requested flag
 
     // Step 9
-    // TODO: handle CORS not set and cross-origin blocked
+    // https://fetch.spec.whatwg.org/#cross-origin-resource-policy-header
+    fn cross_origin_resource_policy_check(request: &Request, response: &Response) -> bool {
+        // Step 1
+        if request.mode != RequestMode::NoCors {
+            return true;
+        }
+
+        // Step 2
+        let current_url_origin = request.current_url().origin();
+        let same_origin = if let Origin::Origin(ref origin) = request.origin {
+            *origin == request.current_url().origin()
+        } else {
+            false
+        };
+
+        if same_origin {
+            return true;
+        }
+
+        // Step 3
+        let policy = response
+            .headers
+            .get(HeaderName::from_static("cross-origin-resource-policy"))
+            .map(|h| h.to_str().unwrap_or(""))
+            .unwrap_or("");
+
+        // Step 4
+        if policy == "same-origin" {
+            return false;
+        }
+
+        // Step 5
+        if let Origin::Origin(ref request_origin) = request.origin {
+            let schemeless_same_origin = if !request_origin.is_tuple() &&
+                !current_url_origin.is_tuple() &&
+                *request_origin == current_url_origin
+            {
+                true
+            } else if request_origin.is_tuple() && current_url_origin.is_tuple() {
+                let request_origin_host = request_origin
+                    .host()
+                    .map(|h| h.to_string())
+                    .unwrap_or("".into());
+                let current_url_origin_host = current_url_origin
+                    .host()
+                    .map(|h| h.to_string())
+                    .unwrap_or("".into());
+                let request_origin_reg = reg_suffix(&request_origin_host);
+                let current_url_reg = reg_suffix(&current_url_origin_host);
+
+                (request_origin.host() == current_url_origin.host() && request_origin_reg == "") ||
+                    (request_origin_reg != "" && request_origin_reg == current_url_reg)
```

Comment:
I had to read this several times closely to realize what was confusing me about it. Let's write the second check in the same order as the first: first compare the registrable domains, then check that one of them is non-null.